storm-core/src/jvm/org/apache/storm/scheduler/IsolationScheduler.java

```diff
@@ -0,0 +1,417 @@
+                        cluster.freeSlot(ass.getWorkerSlot());
+                    }
+                }
+            }
+        }
+
+        Map<String, Set<WorkerSlot>> hostUsedSlots = hostUsedSlots(cluster);
+        LinkedList<HostAssignableSlots> hss = hostAssignableSlots(cluster);
+        List<String> failedTopologyIds = new ArrayList<String>();
+        for (Map.Entry<String, Set<Set<ExecutorDetails>>> entry : topologyWorkerSpecs.entrySet()) {
+            String topologyId = entry.getKey();
+            Set<Set<ExecutorDetails>> executorSet = entry.getValue();
+            if (executorSet != null && executorSet.size() > 0) {
+                failedTopologyIds.add(topologyId);
+            }
+            List<Integer> workerNum = distributionSortedAmts(topologyMachineDistributions.get(topologyId));
+            for (Integer num : workerNum) {
+                HostAssignableSlots hostSlots = hss.peek();
+                List<WorkerSlot> slot = hostSlots != null ? hostSlots.getWorkerSlots() : null;
+
+                if (slot != null && slot.size() >= num.intValue()) {
+                    hss.poll();
+                    cluster.freeSlots(hostUsedSlots.get(hostSlots.getHostName()));
+                    for (WorkerSlot tmpSlot : slot.subList(0, num)) {
+                        Set<ExecutorDetails> executor = removeElemFromSet(executorSet);
+                        cluster.assign(tmpSlot, topologyId, executor);
+                    }
+                    cluster.blacklistHost(hostSlots.getHostName());
+                }
+            }
+        }
+
+        if (failedTopologyIds.size() > 0) {
+            LOG.warn("Unable to isolate topologies " + failedTopologyIds
+                    + ". No machine had enough worker slots to run the remaining workers for these topologies. "
+                    + "Clearing all other resources and will wait for enough resources for "
+                    + "isolated topologies before allocating any other resources.");
+            // clear workers off all hosts that are not blacklisted
+            Map<String, Set<WorkerSlot>> usedSlots = hostUsedSlots(cluster);
+            Set<Map.Entry<String, Set<WorkerSlot>>> entries = usedSlots.entrySet();
+            for (Map.Entry<String, Set<WorkerSlot>> entry : entries) {
+                if (!cluster.isBlacklistedHost(entry.getKey())) {
+                    cluster.freeSlots(entry.getValue());
+                }
+            }
+        } else {
+            // run default scheduler on non-isolated topologies
+            Set<String> allocatedTopologies = allocatedTopologies(topologyWorkerSpecs);
+            Topologies leftOverTopologies = leftoverTopologies(topologies, allocatedTopologies);
+            DefaultScheduler.defaultSchedule(leftOverTopologies, cluster);
+        }
+        cluster.setBlacklistedHosts(origBlacklist);
+    }
+
+    public Set<ExecutorDetails> removeElemFromSet(Set<Set<ExecutorDetails>> executorsSets) {
```

Comment:
Minor: we could make this to have generic implementation (`Set<?> set`) as same as clojure code, or rename to specific to executorsSets and change public to private.